### ISSUE TITLE
perf is potentially being miscompiled

### ISSUE THREAD

**Issue by yshui** [NixOS/nixpkgs#282983]
### Describe the bug
simple commands like `perf top` is segfaulting. the line causing the segfault is `util/dsos.c:160`.

```c
		if (rc == 0) {
			/*
			 * In case the new DSO is a duplicate of an existing
			 * one, print a one-time warning & put the new entry
			 * at the end of the list of duplicates.
			 */
			if (!dso || (dso == this))
				return this;	/* Find matching dso */
			/*
			 * The core kernel DSOs may have duplicated long name.
			 * In this case, the short name should be different.
			 * Comparing the short names to differentiate the DSOs.
			 */
			rc = dso__cmp_short_name(dso, this);  // <--- !!!!!!!!!!!! this line !!!!!!!!!!!!!!!
			if (rc == 0) {
				pr_err("Duplicated dso name: %s\n", name);
				return NULL;
			}
		}
```

it reached line 160 with `dso == NULL`. IOW the `if (!dso ...)` condition is being elided by the compiler for some reason. reduced the optimization level fix the problem. clang seems to compile this fine.

### Metadata
Please run `nix-shell -p nix-info --run "nix-info -m"` and paste the result.

```console
[user@system:~]$ nix-shell -p nix-info --run "nix-info -m"
output here
```

---

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

**Comment by yshui** [comment]
maybe related: #267993 

I am bisecting from the commit listed in that issue.

**Comment by yshui** [comment]
Reduced and reported to upstream: https://gcc.gnu.org/bugzilla/show_bug.cgi?id=113551

In the mean time maybe consider not building perf with gcc13.

**Comment by Ten0** [comment]
What's a valid `override`/`overrideAttrs` to get it to build a version that works?
I tried:
```nix
let gcc12_stdenv = pkgs.overrideCC pkgs.stdenv pkgs.gcc12;
in linuxKernel.packages.linux_6_7.perf.override { stdenv = gcc12_stdenv; }
```
but the resulting derivation still segfaults 😥

**Comment by yshui** [comment]
@Ten0 yeah the perf derivation is weird, it uses the stdenv that was used to build the kernel. so you have to override the kernel package, otherwise perf won't pick that up.

**Comment by yshui** [comment]
(forgot to @viric)

**Comment by Ten0** [comment]
My current workaround:
```nix
let pkgs12 = (import <nixpkgs> { config = { replaceStdenv = ({ pkgs }: pkgs.overrideCC pkgs.stdenv pkgs.gcc12); }; });
in pkgs12.linuxKernel.packages.linux_6_7.perf
```
takes a while to compile, probably sub-optimal, but works.

**Comment by JJJollyjim** [comment]
Here's a workaround which only rebuilds 3 packages:
```nix
(linuxKernel.packagesFor (linuxKernel.kernels.linux_6_7.override {stdenv = gcc12Stdenv; buildPackages = pkgs.buildPackages // { stdenv = gcc12Stdenv;};})).perf
```

**Comment by tobim** [comment]
This is fixed with GCC 13.3 on `nipkgs-unstable`, but 24.05 is still on GCC 13.2. Closing this because of low activity. Please reopen if you need a backport.
